gdnative-core/src/core_types/variant.rs

```diff
@@ -1602,6 +1604,37 @@ impl<T: FromVariant> FromVariant for Vec<T> {
     }
 }
 
+impl<K: ToVariant + Hash + ToVariantEq, V: ToVariant> ToVariant for HashMap<K, V> {
+    #[inline]
+    fn to_variant(&self) -> Variant {
+        let dict = Dictionary::new();
+        for (key, value) in self {
+            dict.insert(key, value);
+        }
+        dict.owned_to_variant()
+    }
+}
+
+impl<K: FromVariant + Hash + Eq, V: FromVariant> FromVariant for HashMap<K, V> {
+    #[inline]
+    fn from_variant(variant: &Variant) -> Result<Self, FromVariantError> {
+        let dictionary = Dictionary::from_variant(variant)?;
+        let len: usize = dictionary
+            .len()
+            .try_into()
+            .expect("Dictionary length should fit in usize");
+        let mut hash_map = HashMap::with_capacity(len);
+        for (key, value) in dictionary.iter() {
+            hash_map.insert(
+                K::from_variant(&key)?,
+                // Maybe add a custom FromVariantError variant if this fails
+                V::from_variant(&value)?,
+            );
```

Comment:
These two `from_variant` calls are actually not that unlikely to fail -- Godot allows homogenous value _and_ key types in dictionaries.

Maybe we could handle errors similarly to  `Vec<T>`:

https://github.com/godot-rust/godot-rust/blob/13417d9858c5daae9462b36564e2c8b94847e72d/gdnative-core/src/core_types/variant.rs#L1594-L1598

Except that I'd probably use this error enumerator:
```rs
InvalidVariantType {
    variant_type: VariantType,
    expected: VariantType,
},
```

Reply:
If the variant type is not the expected by the `from_variant` call, then doesn't it already return a InvalidVariantType ? \
I thought we could do a `map_err` **only** on the value, so that if a value is unexpected, the error also contains the key in the dictionary (just like the `InvalidItem` error contains the index).

Reply:
> If the variant type is not the expected by the `from_variant` call, then doesn't it already return a `InvalidVariantType` ?

True, good point -- maybe add two small test cases, one with an invalid key and one with invalid value? And then `assert_eq!(result, Err(FromVariantError::InvalidVariantType { ... }))`.

Reply:
Done